cmd/query/app/server.go

```diff
@@ -235,17 +239,31 @@ func createHTTPServer(
 
 	apiHandler.RegisterRoutes(r)
 	var handler http.Handler = r
-	handler = responseHeadersHandler(handler, queryOpts.HTTP.ResponseHeaders)
 	if queryOpts.BearerTokenPropagation {
 		handler = bearertoken.PropagationHandler(telset.Logger, handler)
 	}
-	handler = handlers.CompressHandler(handler)
 	recoveryHandler := recoveryhandler.NewRecoveryHandler(telset.Logger, true)
+	handler = recoveryHandler(handler)
+
+	return handler, r
+}
+
+func createHTTPServerLegacy(
+	ctx context.Context,
+	querySvc *querysvc.QueryService,
+	metricsQuerySvc querysvc.MetricsQueryService,
+	queryOpts *QueryOptions,
+	tm *tenancy.Manager,
+	telset telemetery.Setting,
+) (*httpServer, error) {
+	handler, r := createHTTPRouter(querySvc, metricsQuerySvc, queryOpts, tm, telset)
+	handler = responseHeadersHandler(handler, queryOpts.HTTP.ResponseHeaders)
+	handler = handlers.CompressHandler(handler)
```

Comment:
Can we not move all handlers into one factory function?

Also, I thing the recovery handler needs to be the last one, it's catching panics.

Reply:
@yurishkuro we could but OTEL already has functionality for responseHeadersHandler and CompressHandler so we may just end up doing redundant work. Let me know what you think.

Reply:
I changed `createHTTPRouter` to just return the router and inlined the different handlers. Let me know what you think.

Reply:
I am not familiar with what responseHeadersHandler OTEL could provide, ours is rather specific. We should have unit tests that validate both responseHeadersHandler and compression

Reply:
https://github.com/open-telemetry/opentelemetry-collector/blob/3e13ca11e000ed3b5494b4565c1cedbd8e24815e/config/confighttp/confighttp.go#L492

yes that should be fine as responseHeadersHandler. Since we don't always know if unit tests check everything, the way I would be making these changes is (1) remove responseHeadersHandler from main branch and verify that some unit test would break. If so then it's safe to (2) swap that handler for otel helper, since the same test should catch it if something's missing.

Reply:
I'm resolving this thread because for legacy function should not be changed anyway.